mesonbuild/interpreter/interpreter.py

```diff
@@ -2212,67 +2213,68 @@ def func_install_subdir(self, node: mparser.BaseNode, args: T.Tuple[str],
         self.build.install_dirs.append(idir)
         return idir
 
-    @FeatureNewKwargs('configure_file', '0.47.0', ['copy', 'output_format', 'install_mode', 'encoding'])
-    @FeatureNewKwargs('configure_file', '0.46.0', ['format'])
-    @FeatureNewKwargs('configure_file', '0.41.0', ['capture'])
-    @FeatureNewKwargs('configure_file', '0.50.0', ['install'])
-    @FeatureNewKwargs('configure_file', '0.52.0', ['depfile'])
-    @FeatureNewKwargs('configure_file', '0.60.0', ['install_tag'])
-    @permittedKwargs({'input', 'output', 'configuration', 'command', 'copy', 'depfile',
-                      'install_dir', 'install_mode', 'capture', 'install', 'format',
-                      'output_format', 'encoding', 'install_tag'})
     @noPosargs
-    def func_configure_file(self, node, args, kwargs):
-        if 'output' not in kwargs:
-            raise InterpreterException('Required keyword argument "output" not defined.')
-        actions = {'configuration', 'command', 'copy'}.intersection(kwargs.keys())
-        if len(actions) == 0:
+    @typed_kwargs(
+        'configure_file',
+        KwargInfo('output', str, required=True),
+        KwargInfo('capture', bool, default=False, since='0.41.0'),
+        KwargInfo('format', str, default='meson', since='0.46.0',
+                  validator=in_set_validator({'meson', 'cmake', 'cmake@'})),
+        KwargInfo('output_format', str, default='c', since='0.47.0',
+                  validator=in_set_validator({'c', 'nasm'})),
+        DEPFILE_KW.evolve(since='0.52.0'),
+        # Cannot use shared implementation until None backwards compat is dropped
+        KwargInfo('install', (bool, NoneType), since='0.50.0'),
+        KwargInfo('install_dir', (str, bool), default='',
+                  validator=lambda x: 'must be `false` if boolean' if x is True else None),
+        INSTALL_MODE_KW.evolve(since='0.47.0,'),
+        KwargInfo('encoding', str, default='utf-8', since='0.47.0'),
+        KwargInfo('copy', bool, default=False, since='0.47.0'),
+        INSTALL_TAG_KW.evolve(since='0.60.0'),
+        KwargInfo(
+            'command',
+            ContainerTypeInfo(list, (build.Executable, ExternalProgram, compilers.Compiler, mesonlib.File, str)),
+            listify=True,
+            default=[],
+        ),
+        KwargInfo(
+            'input',
+            ContainerTypeInfo(list, (mesonlib.File, str)),
+            listify=True,
+            default=[],
+        ),
+        KwargInfo(
+            'configuration',
+            (ContainerTypeInfo(dict, (str, int, bool)), build.ConfigurationData),
+            default={},
+        ),
```

Comment:
Alright, if we want to error out on empty configurations this should really have a validator that rejects it directly rather than just letting it evaluate to `bool(False)`. Fun fact: `configuration: configuration_data()` has different truthy status than `{}`.

Reply:
That's easy to fix :)

Reply:
Indeed, seems reasonable. :)